go/vt/mysqlctl/s3backupstorage/s3.go

```diff
@@ -427,6 +428,20 @@ func (bs *S3BackupStorage) WithParams(params backupstorage.Params) backupstorage
 	return &S3BackupStorage{params: params}
 }
 
+// This creates a new transport based off http.DefaultTransport the first time and returns the same
+// transport on subsequent calls so connections can be reused as part of the same transport.
+func (bs *S3BackupStorage) getTransport() *http.Transport {
+	if bs.transport == nil {
```

Comment:
Same question as before though, does this get run concurrently? 

I think we can create the transport whenever we create the struct instead to avoid that?

Reply:
it is possible, considering that we will have multiple checks on different backups can happen at the same time. perhaps adding a mutex here would be better so it is always safe, regardless if it is set when we create the struct or when we first call this function, what do you think?

Reply:
actually, this is only called on `S3BackupStorage.client()` which acquires a mutex when it is running, so we shouldn't see this running concurrently

https://github.com/vitessio/vitess/blob/eff2a9276a9cc970e5835a33a505f31fc3d6012f/go/vt/mysqlctl/s3backupstorage/s3.go#L457-L463

Reply:
@rvrangel i think we can then also move the initialization there too? No need to wrap it in another lazy loading function?

Reply:
Sorry, that’s wrong. I mean where `bs` is initialized.

Reply:
yeah, we can do that if you prefer - originally I didn't to make it easier to have a unit test around this (and the existing `client()` function doesn't have a test, looks like it makes a call to S3 to validate it) and since we are not using a contructor function like `New()` to create this struct - everything gets done in this function - it is harder to mock the S3 client that makes an actual `HeadBucket()` call.

so my question is, would you be okay with moving this into `client()` without a unit test, or do you want to make this into a bigger PR where we can mock that call and change how `S3BackupStorage` gets created?

Reply:
there is also the consideration that we might have a single `S3BackupStorage` but we might call `S3BackupStorage.client()` for each request - which then will use a different transport for each client we initialise, while the current approach uses the same transport for all clients generated by the same `S3BackupStorage`

Reply:
I think a small refactor with a (private) constructor and then testing transport is fine?

I don’t see how we create different transports though? We can then set directly `bs.transport`?

Reply:
taking a better look at the code, we do that initialisation under `init()` so it should be easy enough to update it with a `new()` method instead of the current line, so I will update it tomorrow:
https://github.com/vitessio/vitess/blob/eff2a9276a9cc970e5835a33a505f31fc3d6012f/go/vt/mysqlctl/s3backupstorage/s3.go#L512-L513

but if we set `bs.transport` directly, the problem is the functions in this struct were designed in a way to call `bs.client()` every time, like here:
https://github.com/vitessio/vitess/blob/eff2a9276a9cc970e5835a33a505f31fc3d6012f/go/vt/mysqlctl/s3backupstorage/s3.go#L292-L294

as an example take `S3BackupStorage.ListBackups()`, every time we call it, it will in turn call `S3BackupStorage.client()` which will set `bs.transport = ` again, making it a different transport each time. or unless you meant to lazy load it inside the `client()` function as we are doing in the current chage?

Reply:
@dbussink I have moved this to a private constructor as we talked, and instead of setting `bs.transport` inside `client()` it is now only set on the constructor - and I have removed the lazy loading, but that will cause issues if there is any place where we use `S3StorageBackup{}` without initialising the transport, but it doesn't seem like we do?

let me know if this looks good to you or if you want to add extra safety on the `client()` function